include/Architecture.hpp

```diff
@@ -406,9 +416,12 @@ class Architecture {
     fidelityDistanceTables.clear();
   }
 
-  [[nodiscard]] double distance(std::uint16_t control,
-                                std::uint16_t target) const {
+  [[nodiscard]] double distance(std::uint16_t control, std::uint16_t target,
+                                bool includeReversalCost = true) const {
     if (currentTeleportations.empty()) {
+      if (includeReversalCost) {
+        return distanceTableReversals.at(control).at(target);
+      }
```

Comment:
I thought the naming was clear enough here (should I add doc strings for those tables?). `distanceTable` contains just the swap distance (i.e. the cheapest path distance from 1 qubit to another qubit skipping 1 edge, on which the respective gate is executed) and `distanceTableReversals` contains the swap distance including the minimum cost for reversals (currently at most 1 reversal is considered). The former is needed in the new `GateCountSumDistanceMinusSharedSwaps` heuristic, which has to handle gate reversals separately to ensure principal admissibility.

I think, given only purely bi-/unidirectional architectures, the current gate costs and as long as 1 reversal remains the upper limit we consider, we could also just use only `distanceTable` and add COST_DIRECTION_REVERSE if needed, since the cheapest path should not change in that case, but I would argue that the added flexibility is definitely worth the second table (probably not for the unrealistic use-case of semi-directional architectures, but definitely for variable gate costs and arbitrary 2q gates)

Reply:
Alright. With the added context you provided, things become much clearer.
The added flexibility certainly is a plus here. And as you said: it might be necessary for handling arbitrary two-qubit gates.
Space-wise, it's also not really an issue; even if we were to consider 1000+ qubit architectures.
Runtime-wise it's a bit more of an overhead, but I guess it's also fine. (Again, some kind of caching could help in addition)